File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> results;
        vector<int> result;
        
        dfs(0, 0, target, candidates, result, results);
        
        return results;
    }
    
    void dfs(int start, int sum, int target, vector<int>& candidates, vector<int>& result, vector<vector<int>>& results)
    {
        if(sum > target) return;
        if(sum == target)
        {
            results.push_back(result);
            return;
        }
        
        for(int i = start; i < candidates.size(); i++)
        {
            result.push_back(candidates[i]);
            sum += candidates[i];
            dfs(i, sum, target, candidates, result, results);
            sum -= candidates[i];
            result.pop_back();
        }
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        
        
        def recur(start, cur_sum, result, results):
            if cur_sum >= target:
                
                if cur_sum == target:
                    results.append(result[:])
                
                return
            
            
            for i in range(start, len(candidates)):
                cur_sum += candidates[i]
                result.append(candidates[i])
                
                recur(i, cur_sum, result, results)
                cur_sum -= candidates[i]
                result.pop()
        
        result = []
        results = []
        
        recur(0, 0, result, results)
        return results
            